woocommerce/woocommerce-ios
# [Mobile Payments] Add support for fetching status of Stripe Extension

Closes: #5527

### Instructions to Reviewers
- @ctarda - I cannot assign you as a reviewer since you originated this PR, but your feedback is most welcome
- @jaclync - would you have some time to review this? thank you!

### Description
- Fetches the status of the Stripe extension.
- Ensures that both the WooCommerce Payments extension AND the Stripe extension are not active at the same time.
- Fetches the payment gateway accounts behind the WooCommerce Payments extension AND the Stripe extension.

### How This Fits Into the Bigger Picture
- We have recently added the accounts remote for the Stripe extension, although there are some issues with it, so please see the changes suggested below to work around that. See https://github.com/woocommerce/woocommerce-gateway-stripe/pull/2200#issuecomment-1005241765
- That said, with the accounts remote in place, we can now proceed to at least verify that settings can properly handle either plugin being the payment gateway account provider. That's what this PR focuses on.
- This does NOT ensure token fetching, customer creation or payment capture work. Those are separate issues.

### Nuances
- This is not an ideal solution. PaymentGatewayAccountStore::loadAccounts is unaware of whether or not the WooCommerce Payments and Stripe extensions are active and will attempt to fetch account information from both. A better solution, beyond the scope of this PR, would be to only dispatch the appropriate request for the extension that is active. Issue https://github.com/woocommerce/woocommerce-ios/issues/5853 has been opened to fix this.
- The reasonForAnalytics enums mention wcPay and don't take into account when the reason is related to the Stripe extension. Issue https://github.com/woocommerce/woocommerce-ios/issues/5854 has been opened to address this.

### Testing instructions
- Set up your test store by installing the Stripe extension. Do not activate the WooCommerce Payments extension if it is also installed.
- Modify woocommerce-gateway-stripe/includes/admin/class-wc-rest-stripe-account-controller.php so that get_account_summary also returns keys of is_live (false) and test_mode (true) if necessary
- Activate the Stripe extension
- Set up the Stripe extension using developer (test) keys
- Launch the mobile app. In settings (gear), under Experimental Features, ensure that “IPP with Stripe extension” is enabled
- Navigate back up to the root of settings, and tap on In-Person Payments. Ensure the app is able to load your account successfully and you see the “Order, Manage, Manual” cells
- Note that connecting to a reader and capturing a payment using the Stripe extension is 100% untested and not expected to work.
- Switch to a store running just the WooCommerce Payments extension and ensure everything continues to work as expected, including connecting to a reader and capturing a payment.
- Lastly, enable both extensions on a store and ensure on entering settings (gear), In-Person Payments you are scolded for your foolishness.
- Optional: In PaymentGatewayAccountStore::loadAccounts set breakpoints on both upsertStoredAccountInBackground calls and ensure they are fired as appropriate depending on whether or not each extension is active.

Note: In the output window you may see

> 2022-01-10 13:53:36.259330-0800 WooCommerce[3408:2006840] Error fetching WCPay account - it is possible the extension is not installed or inactive

This is expected for sites with only the Stripe extension active and it is the principal side effect of this PR, since PaymentGatewayAccountStore::loadAccounts is unaware of which plugins are active. You’ll see a similar message if only the WooCommerce Payments extension is active.

### Screenshots

When both extensions are active, the following will be displayed (selectPlugin):

![IMG_0289](https://user-images.githubusercontent.com/1595739/148847231-4af38abf-5277-4302-9646-f8d5435d7262.jpg)

---
